File: Jump_to_Algorithm3/week12/SoohyunKim/1027.고층건물.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

function solution(input) {
  const N = +input[0];
  const buildings = input[1].split(' ').map(Number);
  let answer = 0;

  // 기울기 계산
  const slope = (x1, y1, x2, y2) => {
    return (y2 - y1) / (x2 - x1);
  }

  for (let i = 0; i < N; i++) {
    // 현재 건물 좌표
    const x1 = i + 1; 
    const y1 = buildings[i]; 

    // 오른쪽
    let maxRight = null;  // 가장 큰 기울기
    let visibleRight = 0; // 볼 수 있는 빌딩 수

    for (let j = i + 1; j < N; j++) {
      const x2 = j + 1;
      const y2 = buildings[j];
      const slopeRight = slope(x1, y1, x2, y2);

      // 기울기가 커지는 경우 -> 보인다
      if (maxRight === null || maxRight < slopeRight) {
        maxRight = slopeRight;
        visibleRight++;
      }
    }

    // 왼쪽
    let maxLeft = null;
    let visibleLeft = 0;

    for (let k = i - 1; k >= 0; k--) {
      const x2 = k + 1;
      const y2 = buildings[k];
      const slopeLeft = slope(x1, y1, x2, y2);

      if (maxLeft === null || maxLeft > slopeLeft) {
        maxLeft = slopeLeft;
        visibleLeft++;
      }
  }

    if ((visibleLeft + visibleRight) > answer) {
      answer = visibleLeft + visibleRight;
    }
  }
  
  return answer;
}

console.log(solution(input));

File: Jump_to_Algorithm3/week12/HyeonjiLee/1027 고층 건물.py
N = int(input())
arr = list(map(int, input().split()))
answer = 0

for i in range(N):
    cnt = 0  # 보이는 빌딩 수

    # 오른쪽 기울기 비교
    max_right = -1000000001  # 기울기 초기화 (가장 작은 값으로 설정)
    for j in range(i + 1, N):
        x1, y1 = i, arr[i]
        x2, y2 = j, arr[j]

        a = (y2 - y1) / (x2 - x1)  # 기울기 계산

        if a > max_right:  # 이전 기울기보다 더 큰 기울기일 경우만 보인다
            max_right = a
            cnt += 1

    # 왼쪽 기울기 비교
    max_left = 1000000001   # 기울기 초기화 (가장 큰 값으로 설정)
    for j in range(i - 1, -1, -1):
        x1, y1 = i, arr[i]
        x2, y2 = j, arr[j]

        a = (y2 - y1) / (x2 - x1)  # 기울기 계산

        if a < max_left:  # 이전 기울기보다 더 작은 기울기일 경우만 보인다
            max_left = a
            cnt += 1

    answer = max(answer, cnt)

print(answer)
